### PR TITLE
fix(server): use the config.hostname to open server binding

### PR DESCRIPTION
**Pull request by paulRbr**
__Hello!__

I've had the issue to not be able to bind the karma webserver on the local ip in a cloud instance. 
Realising that Karma was not listenning on the configurated hostname, I am submitting a PR for that.

Let me know if this is a constraint that comes from somewhere I didn't get.

Thanks for this wonderful project!

### CODE REVIEW

**Comment by vojtajina** [comment]
@popox why do you need to listen only on a specific interface? What is the use case for it?

@dlecan Yes, it is not a breaking change, but it does make configuration of Karma more complex so I wanna make sure it is worthy.

**Comment by popox** [comment]
@vojtajina my use case is on a public paas instance (openshift online) where I _don't have any rights to bind a specific port on the local ip `0.0.0.0`_ but I have the rights to listen on the public IP available for my instance.

This helped me to be able to run an integration server on a cloud provider and be able to launch Karma to orchestrate my JS tests.

**Comment by maksimr** [comment]
Good use case from https://github.com/karma-runner/karma/pull/1311

> Current karma version doesn't specify the address and in this case node binds to IPv4 address only. So there is no way to bind it to v6 interface. With added option it's now possible to specify v6 address.

**Comment by maksimr** [comment]
What if we will not add new `setting` and simple adopt `hostname` for this purpose.
If hostname the same as default hostname then listen on `0.0.0.0`

/сс @zzo @aymericbeaumet @Dignifiedquire

**Comment by zzo** [comment]
Sgtm is there a pr for just that change somewhere I missed?

**Comment by maksimr** [comment]
@paulRbr what do you think about using `hostname` for this purprose?

@zzo no, only this PR

**Comment by paulRbr** [comment]
@maksimr If we don't add a new configuration option, we'll be more likely to break someone's usage of Karma.

**I.e.** I am thinking of people configuring the `hostname` to let's say `whatever.example.com` which is pointing to the ip `88.77.66.55`. Without noticing they would limit the server listen host to that specific ip address. Therefore, in case they have scripts / dependencies connecting to karma through the local interface `127.0.0.1` where karma is running, they would break and receive a:
 `Failed to connect to 127.0.0.1 port 9876: Connection refused`

If we go for that, they would need to launch karma with the env variable `IP=88.77.66.55` to be feature equivalent.

/cc @zzo @vojtajina

**Comment by maksimr** [comment]
@paulRbr Thanks!

Yeah...  `hostname` is not best choice.
Will be cool add information about `listen` to documentation

**Comment by paulRbr** [comment]
@maksimr PR updated with doc and rebased

File: lib/server.js [hunk 1]
```diff
@@ -45,7 +45,11 @@ var start = function(injector, config, launcher, globalEmitter, preprocess, file
     if (e.code === 'EADDRINUSE') {
       log.warn('Port %d in use', config.port);
       config.port++;
-      webServer.listen(config.port);
+      if (config.listen) {
+        webServer.listen(config.port, config.listen);
+      } else {
+        webServer.listen(config.ports);
```
Comment (just-boris) [inline lib/server.js]: looks like there is should be `config.port`

Comment (paulRbr) [inline lib/server.js]: good catch, thanks

File: lib/constants.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ exports.VERSION = pkg.version;
 
 exports.DEFAULT_PORT = process.env.PORT || 9876;
 exports.DEFAULT_HOSTNAME = process.env.IP || 'localhost';
+exports.DEFAULT_LISTEN_IP = process.env.IP || '0.0.0.0';
```
Comment (paulRbr) [inline lib/constants.js]: I don't really like the magical env variable as discussed in #1186, should we just leave `0.0.0.0` as default value and not read the `IP` env variable?

Comment (Dignifiedquire) [inline lib/constants.js]: How about we use `KARMA_IP`, I think env variables like that should always be scoped.

Comment (Dignifiedquire) [inline lib/constants.js]: Also we need documentation on that option if we add it i

**Comment by Dignifiedquire** [comment]
Except for the inline comment I think this is a good solution, and as soon as this is addressed I think we can merge this.

**Comment by paulRbr** [comment]
@Dignifiedquire done taking your inline comment into account.

**Comment by Dignifiedquire** [comment]
That was quick :) looks good so far, the last thing I'm worried about is that there is no test for this, but the `server.spec.coffee` is a the moment pretty empty :( But if you feel up to it would be great if you add the first test into there

File: lib/server.js [hunk 3]
```diff
@@ -45,7 +45,11 @@ var start = function(injector, config, launcher, globalEmitter, preprocess, file
     if (e.code === 'EADDRINUSE') {
       log.warn('Port %d in use', config.port);
       config.port++;
-      webServer.listen(config.port);
+      if (config.listen) {
```
Comment (maksimr) [inline lib/server.js]: I think now we can remove `if (config.lister) {` and simple write `webServer.lister(config.port, config.lister)` and `webServer.lister(config.port,config.lister, startCallback)`

Thanks!

**Comment by j0tunn** [comment]
Hey guys, what about this PR?

**Comment by zzo** [comment]
Looks good to me - are we waiting for Maksim's comments to be addressed?

**Comment by Dignifiedquire** [comment]
This needs rebasing and updating onto the latest master. Also there should be test for this, in the config.spec for reading the env variable and in the server.spec for it to be used ( there are mow tests in server.spec)

**Comment by arikon** [comment]
@paulRbr Could you rebase and write some tests, please?

**Comment by j0tunn** [comment]
Are there any chances to merge this PR?

**Comment by Dignifiedquire** [comment]
Closing due to inactivity

**Pull request closed without merging**